### ISSUE TITLE
timestamp-macros.h modified CFLAGS/CXXFLAGS leaking into packages

### ISSUE THREAD

**Issue by pullmoll** [voidlinux/void-packages#3765]
When trying to build the `alsa-tools` package there are errors regarding a missing `/builddir/.xbps-fltk/timestamp-macros.h` file. See http://sprunge.us/iigc

FWIW the `/builddir/.xbps-alsa-tools/timestamp-macros.h` file repeats the macro definitions like this:
```
#undef __DATE__
#define __DATE__ "Nov 10 2015"
#undef __TIME__
#define __TIME__ "18:08:26"
#undef __DATETIME__
#define __DATETIME__ "Nov 10 2015 18:08:26"
#undef __DATE__
#define __DATE__ "Nov 10 2015"
#undef __TIME__
#define __TIME__ "18:08:26"
#undef __DATETIME__
#define __DATETIME__ "Nov 10 2015 18:08:26"
```

Similar problems arise trying to build the `ImageMagick` package. There it is `/builddir/.xbps-perl/timestamp-macros.h` which is missing. See http://sprunge.us/CPPU

In case of `ImageMagick` the sequence of `hostdir=...` and `hostdir+=" perl"` in the template is wrong for non-cross builds. Fixing it (moving the `if [ -z "$CROSS_BUILD" ]; then ...` block down) does not fix above problem, though.

**Comment by Gottox** [comment]
>When trying to build the alsa-tools package there are errors regarding a missing /builddir/.xbps-fltk/timestamp-macros.h file. See http://sprunge.us/iigc

I cannot reproduce this issue. Have you tried to xbps-clean before building?

**Comment by pullmoll** [comment]
I even tried with `./xbps-src zap` and `./xbps-src binary-bootstrap`. I'll try with an empty repository, i.e. without my locally built packages.

**Comment by pullmoll** [comment]
Closing because it seems to be an issue which only appears locally.

**Comment by pullmoll** [comment]
Here it is again on the build bots:
http://build.voidlinux.eu/builders/x86_64_builder/builds/21467/steps/shell_3/logs/stdio

**Comment by Gottox** [comment]
Did you have perl in your local binpkgs?

**Comment by pullmoll** [comment]
Yes. I usually do `./xbps-src -N pkg base-system` and then packages I need with `-N pkg <name>`.

Perhaps it has to do with `subpackages+=...` in the template?

**Comment by Gottox** [comment]
can you try this patch?

```diff
diff --git a/common/environment/build/timestamp-macros.sh b/common/environment/build/timestamp-macros.sh
deleted file mode 120000
index da4ca4f..0000000
--- a/common/environment/build/timestamp-macros.sh
+++ /dev/null
@@ -1 +0,0 @@
-../configure/timestamp-macros.sh
\ No newline at end of file
diff --git a/common/environment/configure/timestamp-macros.sh b/common/environment/configure/timestamp-macros.sh
deleted file mode 100644
index b324356..0000000
--- a/common/environment/configure/timestamp-macros.sh
+++ /dev/null
@@ -1,13 +0,0 @@
-if [ -n "$SOURCE_DATE_EPOCH" ]; then
-	CFLAGS+=" -Wno-builtin-macro-redefined -include$XBPS_STATEDIR/timestamp-macros.h"
-	CXXFLAGS+=" -Wno-builtin-macro-redefined -include$XBPS_STATEDIR/timestamp-macros.h"
-	printf "" > $XBPS_STATEDIR/timestamp-macros.h
-	for i in "DATE,%b %d %Y" "TIME,%H:%M:%S" "DATETIME,%b %d %Y %H:%M:%S"; do
-		mcr=${i%%,*}
-		val=$(LC_ALL=C date --date "@$SOURCE_DATE_EPOCH" +"${i#*,}")
-		cat >> $XBPS_STATEDIR/timestamp-macros.h <<EOF
-#undef __${mcr}__
-#define __${mcr}__ "${val}"
-EOF
-	done
-fi
diff --git a/common/hooks/pre-build/03-timestamp-macros.sh b/common/hooks/pre-build/03-timestamp-macros.sh
new file mode 120000
index 0000000..725e30d
--- /dev/null
+++ b/common/hooks/pre-build/03-timestamp-macros.sh
@@ -0,0 +1 @@
+../pre-configure/03-timestamp-macros.sh
\ No newline at end of file
diff --git a/common/hooks/pre-configure/03-timestamp-macros.sh b/common/hooks/pre-configure/03-timestamp-macros.sh
new file mode 100644
index 0000000..46a8787
--- /dev/null
+++ b/common/hooks/pre-configure/03-timestamp-macros.sh
@@ -0,0 +1,18 @@
+# This hook overwrites timestamp macros.
+#
+
+hook() {
+	if [ -n "$SOURCE_DATE_EPOCH" ]; then
+		CFLAGS+=" -Wno-builtin-macro-redefined -include$XBPS_STATEDIR/timestamp-macros.h"
+		CXXFLAGS+=" -Wno-builtin-macro-redefined -include$XBPS_STATEDIR/timestamp-macros.h"
+		printf "" > $XBPS_STATEDIR/timestamp-macros.h
+		for i in "DATE,%b %d %Y" "TIME,%H:%M:%S" "DATETIME,%b %d %Y %H:%M:%S"; do
+			mcr=${i%%,*}
+			val=$(LC_ALL=C date --date "@$SOURCE_DATE_EPOCH" +"${i#*,}")
+			cat >> $XBPS_STATEDIR/timestamp-macros.h <<EOF
+#undef __${mcr}__
+#define __${mcr}__ "${val}"
+EOF
+		done
+	fi
+}
diff --git a/common/hooks/pre-install/03-timestamp-macros.sh b/common/hooks/pre-install/03-timestamp-macros.sh
new file mode 120000
index 0000000..725e30d
--- /dev/null
+++ b/common/hooks/pre-install/03-timestamp-macros.sh
@@ -0,0 +1 @@
+../pre-configure/03-timestamp-macros.sh
\ No newline at end of file
```

**Comment by Gottox** [comment]
I'm still unable to reproduce the issue here.

**Comment by pullmoll** [comment]
That will take some more time. I just finished building base-system and am now trying to build ImageMagick again. If the problem is still there, I'll give your patch a try.

**Comment by Gottox** [comment]
Okay. Then this code should be dead anyway. So this could be the issue

Nevertheless, the builder still throughs an error with that var unset. So we've got at least two bugs here.

**Comment by pullmoll** [comment]
Are you saying `XBPS_USE_BUILD_MTIME` is known or supposed to fail?

If you look at http://sprunge.us/iigc, you will see there is a `-include/builddir/.xbps-<something>/timestamp-macros.h` in the `CFLAGS` **twice**. I think this should never happen, because the second include would redefine the same set of macros.

So IMO the question is why it happens and how to avoid it.

**Comment by Gottox** [comment]
> Are you saying XBPS_USE_BUILD_MTIME is known or supposed to fail?
No, but the compiler flags should never be setted if it is defined.

> Simply scan for timestamp-macros.h in the current CFLAGS and skip adding another one in case there already is one?
I'm not sure if that would work. There are two calls to this script, one for an dependency one for the package itself. If the wrong one is called first, it would lead to the same error. The dependency call should never be called in that context, imo.

@xtraeme is this behavior expected?

**Comment by pullmoll** [comment]
Then the `timestamp-macros.sh` script should perhaps check if `${XBPS_STATEDIR#${XBPS_BUILDDIR}/.xbps-} == ${XBPS_PKG}` or something along that line.
In other words it could check if it is called for the package itself, not for a dependency.

**Comment by pullmoll** [comment]
> > Are you saying XBPS_USE_BUILD_MTIME is known or supposed to fail?

> No, but the compiler flags should never be setted if it is defined.

Then the script should also check for `-z "$XBPS_USE_BUILD_MTIME"`.
As it is now it looks like `$SOURCE_DATE_EPOCH` is defined, for whatever reason, even if `$XBPS_USE_BUILD_MTIME` is set.

**Comment by pullmoll** [comment]
Ok, so now I tested removing `common/environment/configure/timestamp-macros.sh` and instead creating `common/hooks/pre-configure/03-timestamp-macros.sh` with the following content:

```
# This hook overwrites timestamp macros.
#
hook() {
	local i mcr val macros="$XBPS_STATEDIR/timestamp-macros.h"

	[ -n "$XBPS_USE_BUILD_MTIME" ] && return 0
	[ -z "$SOURCE_DATE_EPOCH" ] && return 0
	[ "${XBPS_STATEDIR#${XBPS_BUILDDIR}/.xbps-}" != "${PKGNAME}" ] && return 0
	msg_normal "Creating $macros\n"
	CFLAGS+=" -Wno-builtin-macro-redefined -include$macros"
	CXXFLAGS+=" -Wno-builtin-macro-redefined -include$macros"
	rm -f "$macros"
	for i in "DATE,%b %d %Y" "TIME,%H:%M:%S" "DATETIME,%b %d %Y %H:%M:%S"; do
		mcr=${i%%,*}
		val=$(LC_ALL=C date --date "@$SOURCE_DATE_EPOCH" +"${i#*,}")
		echo "#undef __${mcr}__" >> "$macros"
		echo "#define __${mcr}__ \"${val}\"" >> "$macros"
	done
}
```

When building with `./xbps-src pkg alsa-tools` this script is sourced once and the msg_normal output is `=> Creating /builddir/.xbps-alsa-tools/timestamp-macros.h`.

Yet still, when building the subpackage `hdspconf`, there is another `-include/builddir/.xbps-fltk/timestamp-macros.h` for the `fltk` dependency in the `CXXFLAGS`.

The compiler call looks like this:
```
g++ -DPACKAGE_NAME=\"hdspconf\" -DPACKAGE_TARNAME=\"hdspconf\" -DPACKAGE_VERSION=\"1.4\" -DPACKAGE_STRING=\"
hdspconf\ 1.4\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DPACKAGE=\"hdspconf\" -DVERSION=\"1.4\" -DSTDC_
HEADERS=1 -DHAVE_LIBASOUND=1 -I.     -specs=/void-packages/common/environment/configure/gccspecs/hardened-cc
1 -fstack-protector-strong -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe    -Wno-builtin-macro-redefined -inc
lude/builddir/.xbps-alsa-tools/timestamp-macros.h  -I/usr/include/freetype2 -specs=/void-packages/common/env
ironment/configure/gccspecs/hardened-cc1 -fstack-protector-strong -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pi
pe -Wno-builtin-macro-redefined -include/builddir/.xbps-fltk/timestamp-macros.h -fvisibility-inlines-hidden
-D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -D_THREAD_SAFE -D_REENTRANT -fno-exceptions
 -MT hdspconf.o -MD -MP -MF .deps/hdspconf.Tpo -c -o hdspconf.o hdspconf.cxx
```

The reason for this unwanted part in the `$CXXFLAGS` is  **a leakage of `$CFLAGS` and/or `$CXXFLAGS` when building the `fltk-devel` package**: `fltk-config` contains two lines
```
CFLAGS="-I/usr/include/freetype2 -specs=/void-packages/common/environment/configure/gccspecs/hardened-cc1  -
fstack-protector-strong -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe    -Wno-builtin-macro-redefined -includ
e/builddir/.xbps-fltk/timestamp-macros.h  -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -
D_THREAD_SAFE -D_REENTRANT"
CXXFLAGS="-I/usr/include/freetype2 -specs=/void-packages/common/environment/configure/gccspecs/hardened-cc1
-fstack-protector-strong -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe    -Wno-builtin-macro-redefined -inclu
de/builddir/.xbps-fltk/timestamp-macros.h -fvisibility-inlines-hidden  -D_LARGEFILE_SOURCE -D_LARGEFILE64_SO
URCE -D_FILE_OFFSET_BITS=64 -D_THREAD_SAFE -D_REENTRANT"
```

Now we have to find a way how to avoid these flags from leaking into package config files!

**Comment by pullmoll** [comment]
Here's my suggestion for a `common/hooks/post-install/07-cleanup-timestamp-macros.sh` file:
```
# Remove " -Wno-builtin-macro-redefined" and
# " -include/builddir/.xbps-<pkgname>/timestamp-macros.h"
# from shell scripts in /usr/bin and package config
# files in /usr/lib/pkgconfig

clean_file() {
	msg_normal "Cleaning $1...\n"
	sed -i "$1" \
		-e "s; -Wno-builtin-macro-redefined;;" \
		-e "s; -include${XBPS_BUILDDIR}/\.xbps-.*/timestamp-macros\.h;;"
}

hook() {
	local f mimetype

	if [ -d "${PKGDESTDIR}/usr/bin" ]; then
		# Clean up shell scripts
		for f in "${PKGDESTDIR}/usr/bin/"*; do
			mimetype=$(file --mime-type "$f" | sed "s;$f: ;;")
			if [ "$mimetype" == "text/x-shellscript" ]; then
				clean_file "$f"
			fi
		done
	fi
	if [ -d "${PKGDESTDIR}/usr/lib/pkgconfig" ]; then
		# Clean up package config files
		for f in ${PKGDESTDIR}/usr/lib/pkgconfig/*; do
			clean_file "$f"
		done
	fi
}
```

After rebuilding `fltk-devel` with this post-processing installed, the `fltk-config` file is clean and I can finally build `alsa-tools` again.

**Comment by pullmoll** [comment]
Another thought on this matter: Instead of adding ` -Wno-builtin-macro-redefined` it should be possible to write the line `#pragma GCC diagnostic ignored "-Wbuiltin-macro-redefined"` at the top of the `timestamp-macros.h` file. This pragma could even be pushed/popped to avoid ignoring redefinitions elsewhere in the source.

**Comment by pullmoll** [comment]
With the postprocessing step there are more packages revealed where the timestamp-macro additions leak into the resulting package:

+ Package `ImageMagick`
```
=> WARNING: Can't clean /usr/lib/ImageMagick-6.9.3/config-Q16/configure.xml (mime-type: application/xml) ...
```
+ Package `postgresql` subpackage `postgresql-libs-devel`
```
=> WARNING: Can't clean /usr/lib/postgresql/pgxs/src/Makefile.global (mime-type: text/x-makefile) ...
=> WARNING: Can't clean /usr/bin/pg_config (mime-type: application/x-sharedlib) ...
```

I think pg_config is one of the examples where we have to disable the timestamp-macros, because I see no way to patch applications.

**Comment by pullmoll** [comment]
Should be solved with #3787
